### PR TITLE
✨ [source-datadog] refactor connector and add new stream `service_level_objectives_history`

### PR DESCRIPTION
**Pull request by dlecocq**
## What

At least of the options in the DataDog connector were overloaded to be used across multiple streams. In some cases, this was appropriate (for example, the maximum number of records retrieved in a single request), and in some cases not (for example, the "`query`", which may not be appropriate to use for _both_ SLOs and logs (and other streams). In service of that, stream-specific configuration has been isolated to their own configuration groups, and imposing a naming convention (`<stream>__<setting>`) to differentiate `service_level_objectives__query` and `logs__query`.

In addition, not all of the options available on the DataDog endpoints were previously exposed. This MR adds several (for example, `monitors__tags`, `monitors__with_downtimes`, etc.), though even this change is not exhaustive. My _hope_ is that if this convention suits the community, we would expand it to the remaining endpoints so that Airbyte exposes more to the connector user.

Lastly, this adds a new stream, the `service level objectives history`.

## How
(Captured above)

## 🚨 User Impact 🚨
**Breaking changes** - some of the existing configuration options have been supplanted by their newer versions:

- `query` -> `logs__query`
- `queries` -> removed (it was not referenced anywhere in the manifest)

**Version bump** - an argument could be made for a major bump, but I would suggest a minor version bump to `0.5.0`.

### Community member or Airbyter

- ✅ **Community member?** Grant edit access to maintainers
- 🔴 **Unit & integration tests added and passing** I have not been able to get `airbyte-ci` to handle this connector well to run the tests, but I'm working on it. In the meantime, there's value in submitting the MR.

I'll also include some screenshots of it working in my local Airflow:

Screenshot 2024-02-16 at 11 54 13 AM
Screenshot 2024-02-16 at 11 54 20 AM
Screenshot 2024-02-16 at 11 54 27 AM

### CODE REVIEW

**Comment by dlecocq** [comment]
Also worth noting that this depends on #35341

**Comment by natikgadzhi** [comment]
> queries -> removed (it was not referenced anywhere in the manifest)

🤦🏼 Thank you for cleaning it up. We're taking a look.

File: airbyte-integrations/connectors/source-datadog/source_datadog/manifest.yaml [hunk 1]
```diff
@@ -137,13 +137,35 @@ definitions:
...
+      paginator:
+        type: "DefaultPaginator"
+        page_size_option:
+          type: "RequestOption"
+          inject_into: "request_parameter"
+          field_name: "page_size"
+        pagination_strategy:
+          type: "PageIncrement"
+          page_size: "{{ config['max_records_per_request'] }}"
+          start_from_page: 0
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-datadog/source_datadog/manifest.yaml:149]: If that was a string value, would the dependency PR still be required @dlecocq?

Comment (dlecocq) [reply airbyte-integrations/connectors/source-datadog/source_datadog/manifest.yaml:149]: If we set it to an int of our choosing, then they would be decoupled. I did try with `"500"` and the like, and it still complains. It has to be a literal yaml int in order for it to work.

FWIW, this is not a hill that I'd die on - if we wanted to hard code it for now, I think that's a sensible approach.

**Review by natikgadzhi — COMMENTED** [review]
This looks good to me at first glance. Let's merge the CDK one first, then get back to this, should be quick.

File: airbyte-integrations/connectors/source-datadog/source_datadog/manifest.yaml [hunk 2]
```diff
@@ -243,20 +335,206 @@ streams:
...
     - "downtimes"
-    - "incident_teams"
     - "incidents"
-    - "logs"
-    - "metrics"
     - "monitors"
     - "service_level_objectives"
     - "synthetic_tests"
     - "users"
+
```
[the remark this answers is not in the document]
Comment (dlecocq) [reply airbyte-integrations/connectors/source-datadog/source_datadog/manifest.yaml:352]: np - I wasn't sure if that was the convention or not, but to run `read` command line (which is how I did most of my iterating), it seemed to need it all in the manifest.

**Comment by natikgadzhi** [comment]
Once the CDK change gets merged in, let's run tests on this, and merge this.

@dlecocq, if you have a minute, would you be open to improving a few other things in `source-datadog`? Would it be okay if I reach out to you with those changes in a separate issue?

**Comment by natikgadzhi** [comment]
@marcosmarxm heads up, this is _very_ behind master. It might be easier to cherry-pick the changes and reapply them in a fresh PR, if @dlecocq has time.

**Comment by natikgadzhi** [comment]
Ah no I'm wrong, it's actually not bad. But, we'd need to carefully resolve the conflicts.

**Pull request closed without merging**